File: go/leet_code/1700.go
// Package leet_code provides solutions for LeetCode problems.
//
// This file provides a solution for the problem of counting the number of students that cannot eat lunch.
// The students are in a queue and each student either prefers square sandwiches or circular sandwiches.
// The sandwiches are stacked and the student at the front of the queue takes the top sandwich if it is of the type they prefer, otherwise they go to the back of the queue.
// The process continues until no student can eat the top sandwich.
package leet_code

// countStudents counts the number of students that cannot eat lunch.
//
// It accepts two parameters:
// - students: a slice of integers where each integer is 0 if the student prefers square sandwiches and 1 if the student prefers circular sandwiches.
// - sandwiches: a slice of integers where each integer is 0 if the sandwich is square and 1 if the sandwich is circular.
//
// The function returns an integer representing the number of students that cannot eat lunch.
//
// The function uses a counter to keep track of the number of students that prefer each type of sandwich.
// It iterates over the sandwiches, and for each sandwich, if there is a student that prefers the type of the sandwich, it decrements the counter of that type, otherwise it breaks the loop.
//
// Time complexity analysis:
// - Best-case: O(n), when all students prefer the type of the top sandwich.
// - Worst-case: O(n), when all students prefer a type of sandwich that is not at the top.
// - Average-case: O(n), as we always have to iterate over each sandwich in the worst case.
func countStudents(students []int, sandwiches []int) int {
	var stuCount = make([]int, 2)
	for _, stu := range students {
		stuCount[stu]++
	}

	for _, sandwich := range sandwiches {
		if stuCount[sandwich] == 0 {
			break
		}
		stuCount[sandwich]--
	}

	return stuCount[0] + stuCount[1]
}

File: java/leet_code/1700.java
/**
 * This class provides a solution for the problem of counting the number of students that cannot eat
 * a sandwich.
 * The problem is solved by counting the number of students that prefer each type of sandwich and
 * then iterating over the sandwiches.
 * If the current sandwich is preferred by any student, the count of students that prefer this type
 * of sandwich is decreased.
 * If the current sandwich is not preferred by any student, the iteration is stopped.
 * The count of students that cannot eat a sandwich is the sum of the counts of students that prefer
 * each type of sandwich.
 */
class Solution {

    /**
     * Counts the number of students that cannot eat a sandwich.
     *
     * @param students   An array of integers representing the preference of each student. Each
     *                   integer is either 0 or 1, where 0 means the student prefers the type 0
     *                   sandwich and 1 means the student prefers the type 1 sandwich.
     * @param sandwiches An array of integers representing the type of each sandwich. Each integer
     *                   is either 0 or 1, where 0 means the sandwich is of type 0 and 1 means the
     *                   sandwich is of type 1.
     *
     * @return The number of students that cannot eat a sandwich. It is a non-negative integer.
     */
    public int countStudents(final int[] students, final int[] sandwiches) {
        // Initialize an array to count the number of students that prefer each type of sandwich
        final int[] count = new int[2];

        // Count the number of students that prefer each type of sandwich
        for (final int student : students) {
            count[student]++;
        }

        // Iterate over the sandwiches
        for (final int sandwich : sandwiches) {
            // If the current sandwich is not preferred by any student, stop the iteration
            if (0 == count[sandwich]) {
                break;
            }

            // Decrease the count of students that prefer the current sandwich
            count[sandwich]--;
        }

        // Return the sum of the counts of students that prefer each type of sandwich
        return count[0] + count[1];
    }
}
